File: packages/bun-plugin/src/context-stable-ids.ts
import type MagicString from 'magic-string';
import { ts } from 'ts-morph';
import type { SourceFile } from 'ts-morph';

/**
 * Inject stable IDs into createContext() calls for HMR support.
 *
 * Detects `const X = createContext(...)` patterns and injects a `__stableId`
 * argument so the context registry survives bundle re-evaluation.
 * The ID format is `filePath::varName` — unique because file paths are unique
 * and variable names are unique within a file.
 */
export function injectContextStableIds(
  source: MagicString,
  sourceFile: SourceFile,
  relFilePath: string,
): void {
  for (const stmt of sourceFile.getStatements()) {
    if (!ts.isVariableStatement(stmt.compilerNode)) continue;
    for (const decl of stmt.compilerNode.declarationList.declarations) {
      if (!decl.initializer || !ts.isCallExpression(decl.initializer)) continue;
      const callText = decl.initializer.expression.getText(sourceFile.compilerNode);
      if (callText !== 'createContext') continue;
      if (!ts.isIdentifier(decl.name)) continue;

      const varName = decl.name.text;
      const stableId = `${relFilePath}::${varName}`;
      const callExpr = decl.initializer;

      // Insert the stable ID as the last argument
      const argsArr = callExpr.arguments;
      if (argsArr.length === 0) {
        // createContext<T>() → createContext<T>(undefined, 'id')
        const closeParenPos = callExpr.end - 1;
        source.appendLeft(closeParenPos, `undefined, '${stableId}'`);
      } else {
        // createContext<T>(defaultValue) → createContext<T>(defaultValue, 'id')
        const lastArg = argsArr[argsArr.length - 1]!;
        source.appendLeft(lastArg.end, `, '${stableId}'`);
      }
    }
  }
}

File: native/vertz-compiler/src/context_stable_ids.rs
use oxc_ast::ast::{
    BindingPattern, Expression, Statement, VariableDeclarationKind,
};
use oxc_ast::ast::Program;
use oxc_span::GetSpan;

use crate::magic_string::MagicString;

/// Inject stable IDs into `createContext()` calls for HMR support.
///
/// Detects `const X = createContext(...)` patterns at module level and injects
/// a `__stableId` argument so the context registry survives bundle re-evaluation.
/// The ID format is `filePath::varName`.
pub fn inject_context_stable_ids(ms: &mut MagicString, program: &Program, rel_file_path: &str) {
    for stmt in &program.body {
        let Statement::VariableDeclaration(var_decl) = stmt else {
            continue;
        };

        // Only const declarations at module level
        if var_decl.kind != VariableDeclarationKind::Const {
            continue;
        }

        for declarator in &var_decl.declarations {
            // Must have an initializer that is a call expression
            let Some(init) = &declarator.init else {
                continue;
            };
            let Expression::CallExpression(call_expr) = init else {
                continue;
            };

            // Callee must be `createContext`
            let Expression::Identifier(callee) = &call_expr.callee else {
                continue;
            };
            if callee.name.as_str() != "createContext" {
                continue;
            }

            // Binding must be a simple identifier
            let BindingPattern::BindingIdentifier(binding) = &declarator.id else {
                continue;
            };

            let var_name = binding.name.as_str();
            let escaped_path = rel_file_path.replace('\'', "\\'").replace('\\', "\\\\");
            let stable_id = format!("{escaped_path}::{var_name}");

            let args = &call_expr.arguments;
            if args.is_empty() {
                // createContext<T>() → createContext<T>(undefined, 'id')
                // Insert before the closing paren
                let close_paren = call_expr.span.end - 1;
                ms.prepend_left(close_paren, &format!("undefined, '{stable_id}'"));
            } else {
                // createContext<T>(defaultValue) → createContext<T>(defaultValue, 'id')
                let last_arg = &args[args.len() - 1];
                let last_arg_end = last_arg.span().end;
                ms.append_right(last_arg_end, &format!(", '{stable_id}'"));
            }
        }
    }
}
